src/async_client.rs

```diff
@@ -565,9 +562,8 @@ impl AsyncClient {
         let mut response = self.send(request).await?;
 
         for (room_id, room) in &mut response.rooms.join {
-            let mut client = self.base_client.write().await;
```

Comment:
It is okay that I moved this into the for loops right, as soon as I did that the example started working.

Reply:
I think it is, otherwise the client will remain locked until all events are emitted and in a single threaded runtime you can end up in a deadlock if an event callback tries to get a hold of the lock.